tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/WriteSourceRepoProperties.cs

```diff
@@ -0,0 +1,307 @@
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Tasks;
+using Microsoft.Build.Utilities;
+using Microsoft.DotNet.SourceBuild.Tasks.Models;
+using NuGet.Versioning;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Xml.Serialization;
+
+namespace Microsoft.DotNet.Build.Tasks
+{
+    public class WriteSourceRepoProperties : Task
+    {
+        [Required]
+        public string VersionDetailsFile { get; set; }
+
+        [Required]
+        public string ClonedSubmoduleGitRootDirectory { get; set; }
+
+        [Required]
+        public string ClonedSubmoduleDirectory { get; set; }
+
+        [Required]
+        public string SourceBuildMetadataDir { get; set; }
+
+        public override bool Execute()
+        {
+            var serializer = new XmlSerializer(typeof(VersionDetails));
+            VersionDetails versionDetails = null;
+            using (var stream = File.OpenRead(VersionDetailsFile))
+            {
+                versionDetails = (VersionDetails)serializer.Deserialize(stream);
+            }
+
+            var allRepoProps = new Dictionary<string, string>();
+
+            foreach (var dep in versionDetails.ToolsetDependencies.Concat(versionDetails.ProductDependencies))
+            {
+                Log.LogMessage(MessageImportance.Normal, $"[{DateTimeOffset.Now}] Starting dependency {dep.ToString()}");
+                string repoPath = DeriveRepoPath(ClonedSubmoduleDirectory, dep.Uri, dep.Sha);
+                string repoGitDir = DeriveRepoGitDirPath(ClonedSubmoduleGitRootDirectory, dep.Uri);
+                string repoName = GetRepoNameOrDefault(dep);
+                string safeRepoName = repoName.Replace("-", "");
+                try
+                {
+                    WriteMinimalMetadata(repoPath, dep.Uri, dep.Sha);
+                    WriteSourceBuildMetadata(SourceBuildMetadataDir, repoGitDir, dep);
+                    if (File.Exists(Path.Combine(repoPath, ".gitmodules")))
+                    {
+                        HandleSubmodules(repoPath, repoGitDir, dep);
+                    }
+                    allRepoProps[$"{safeRepoName}GitCommitHash"] = dep.Sha;
+                    allRepoProps[$"{safeRepoName}OutputPackageVersion"] = dep.Version;
+                }
+                catch (Exception e)
+                {
+                    Log.LogErrorFromException(e, true, true, null);
+                }
+            }
+
+            string allRepoPropsPath = Path.Combine(SourceBuildMetadataDir, "AllRepoVersions.props");
+            Log.LogMessage(MessageImportance.Normal, $"[{DateTimeOffset.Now}] Writing all repo versions to {allRepoPropsPath}");
+            File.WriteAllText(allRepoPropsPath, allRepoProps.ToString());
+            WritePropsFile(allRepoPropsPath, allRepoProps);
```

Comment:
Won't this overwrite the `File.WriteAllText`? I'm guessing the WriteAllText isn't meant to be here because `allRepoProps.ToString()` doesn't make sense for a dict.

Reply:
Yup, unintentional, thanks.  Fixed.